refactor: Replace fs::absolute calls with AbsPathJoin calls

kiminuo opened the pull request:
This adds better test coverage and will make it easier in #20744 to remove our dependency on the two-argument boost::filesystem::absolute() function which does not have a direct equivalent in C++17.

This PR doesn't change behavior aside from adding an assert and fixing a test bug.

kiminuo committed:
Replace fs::absolute calls with AbsPathJoin calls

This adds better test coverage and will make it easier in #20744 to remove our dependency on the two-argument boost::filesystem::absolute() function which does not have a direct equivalent in C++17.

kiminuo commented on src/fs.h:
```diff
@@ -21,6 +21,15 @@ namespace fs = boost::filesystem;
+     *
+     * @param[in] base  Absolute path.
+     * @param[in] p     Path to combine with \p base.
+     * @return \p p if \p p is an absolute path, othwerwise \p base joined with \p p is returned.
```
```suggestion
     * @return \p p if \p p is an absolute path, otherwise \p base joined with \p p is returned.
```

ryanofsky commented on src/fs.h:
```diff
@@ -21,6 +21,15 @@ namespace fs = boost::filesystem;
     FILE *fopen(const fs::path& p, const char *mode);
 
+    /**
+     * Helper function asserting that \p base is an absolute path and returning the result of boost::filesystem::absolute(p, base) then.
```
In commit "Introduce fsbridge::AbsPathJoin(base, p)." (c0da7456833fac6d5613d97e09266cf54f13594f)

Would be good to drop "and returning the result of boost::filesystem...". Return value is already described below and the point of this function is to avoid the need for people to understand the extremely complicated boost function, which doesn't make any assumptions, by replacing with it a more straightforward join assuming at least one absolute path.

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -3780,7 +3780,7 @@ std::unique_ptr<WalletDatabase> MakeWalletDatabase(const std::string& name, cons
     // 3. Path to a symlink to a directory.
     // 4. For backwards compatibility, the name of a data file in -walletdir.
-    const fs::path& wallet_path = fs::absolute(name, GetWalletDir());
+    const fs::path& wallet_path = fsbridge::AbsPathJoin(GetWalletDir(), name);
```
```suggestion
    const fs::path wallet_path = fsbridge::AbsPathJoin(GetWalletDir(), name);
```

kiminuo replied:
I understand that one is supposed to apply the commit locally without using Github's "Commit suggestion" as that would lead to a new CI run, right?

jonatack replied:
@kiminuo Yes, I don't recall seeing someone use the GitHub web "commit suggestion" interface here other than for just giving feedback, but you may be right. It's probably best (and more secure) to use git locally and not become dependent on GitHub-specific workflows.

jonatack commented on src/wallet/wallettool.cpp:
```diff
@@ -105,7 +105,7 @@ static void WalletShowInfo(CWallet* wallet_instance)
 bool ExecuteWalletToolFunc(const ArgsManager& args, const std::string& command, const std::string& name)
 {
-    fs::path path = fs::absolute(name, GetWalletDir());
+    fs::path path = fsbridge::AbsPathJoin(GetWalletDir(), name);
```
```suggestion
    const fs::path path = fsbridge::AbsPathJoin(GetWalletDir(), name);
```

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -2399,10 +2399,10 @@ static RPCHelpMan dumptxoutset()
         [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
 {
-    fs::path path = fs::absolute(request.params[0].get_str(), GetDataDir());
+    fs::path path = fsbridge::AbsPathJoin(GetDataDir(), request.params[0].get_str());
```
```suggestion
    const fs::path path = fsbridge::AbsPathJoin(GetDataDir(), request.params[0].get_str());
```

kiminuo replied:
Thanks

jonatack commented on src/rpc/blockchain.cpp:
```diff
@@ -2399,10 +2399,10 @@ static RPCHelpMan dumptxoutset()
     // Write to a temporary path and then move into `path` on completion
     // to avoid confusion due to an interruption.
-    fs::path temppath = fs::absolute(request.params[0].get_str() + ".incomplete", GetDataDir());
+    fs::path temppath = fsbridge::AbsPathJoin(GetDataDir(), request.params[0].get_str() + ".incomplete");
```
```suggestion
    const fs::path temppath = fsbridge::AbsPathJoin(GetDataDir(), request.params[0].get_str() + ".incomplete");
```

kiminuo replied:
Thanks

jonatack reviewed (commented):
ACK c0da745683

The PR title should probably be prefixed with "refactor:" -> Edit: I see you just did it :)

A few suggestions below, happy to re-review if you take them.

ryanofsky commented on src/test/fs_tests.cpp:
```diff
@@ -52,6 +52,18 @@ BOOST_AUTO_TEST_CASE(fsbridge_fstream)
         BOOST_CHECK_EQUAL(input_buffer, "bitcoin");
     }
+    {
+        // Join an absolute path and a relative path.
```
In commit "Introduce fsbridge::AbsPathJoin(base, p)." (c0da7456833fac6d5613d97e09266cf54f13594f)

It would be good to add another test joining an absolute path with the empty `""` path and ensuring the absolute path is returned unchanged (no trailing slash). This is the difference between boost and std join functions and cause of bugs in #20744

ryanofsky reviewed (commented):
Conditional code review ACK c0da7456833fac6d5613d97e09266cf54f13594f if appveyor [CreateWallet](https://github.com/bitcoin/bitcoin/blob/f91587f050d9dceb45fe10129a76a4a9a060a09c/src/wallet/test/wallet_tests.cpp#L696) test is fixed:

https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/37264180#L1102

```
C:\projects\bitcoin\src\wallet\test\wallet_tests.cpp(696): Entering test case "CreateWallet"
Assertion failed: base.is_absolute(), file C:\projects\bitcoin\src\fs.cpp, line 36
Command exited with code -1073740791
```

I think it probably needs some extra test setup (maybe a call to GetDataDir()) to avoid the failure.

ryanofsky commented:
I'm able to reproduce appveyor failure on linux with

```
make -C src test/test_bitcoin && src/test/test_bitcoin -l test_suite -t init_tests -t wallet_tests/CreateWallet --random=2
```

To reproduce, it may be necessary to vary the random seed to get CreateWallet test to run after the init_tests as happens on appveyor.

The problem seems to be that init_tests [force set](https://github.com/bitcoin/bitcoin/blob/f91587f050d9dceb45fe10129a76a4a9a060a09c/src/wallet/test/init_test_fixture.cpp#L45) a -walletdir that [GetWalletDir](https://github.com/bitcoin/bitcoin/blob/f91587f050d9dceb45fe10129a76a4a9a060a09c/src/wallet/walletutil.cpp#L16-L19) rejects and turns into an empty path, which gets rejected by the assert because it is not absolute.

Probably a minimal fix is possible in init_tests. Backtrace at assert fail looks like

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff59b38b1 in __GI_abort () at abort.c:79
#2  0x00007ffff59a342a in __assert_fail_base (fmt=0x7ffff5b2aa38 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=assertion@entry=0x5555566fc54f "base.is_absolute()", file=file@entry=0x5555566fc548 "fs.cpp", line=line@entry=36, 
    function=function@entry=0x5555566fc600 <fsbridge::AbsPathJoin(boost::filesystem::path const&, boost::filesystem::path const&)::__PRETTY_FUNCTION__> "boost::filesystem::path fsbridge::AbsPathJoin(const boost::filesystem::path&, const boost::filesystem::path&)") at assert.c:92
#3  0x00007ffff59a34a2 in __GI___assert_fail (assertion=0x5555566fc54f "base.is_absolute()", file=0x5555566fc548 "fs.cpp", line=36, 
    function=0x5555566fc600 <fsbridge::AbsPathJoin(boost::filesystem::path const&, boost::filesystem::path const&)::__PRETTY_FUNCTION__> "boost::filesystem::path fsbridge::AbsPathJoin(const boost::filesystem::path&, const boost::filesystem::path&)") at assert.c:101
#4  0x000055555620de2f in fsbridge::AbsPathJoin (base=..., p=...) at fs.cpp:36
#5  0x00005555563d5298 in MakeWalletDatabase (name="", options=..., status=@0x7fffffffb57c: 21845, error_string=...) at wallet/wallet.cpp:3783
#6  0x0000555555cd0d59 in wallet_tests::TestLoadWallet (chain=...) at wallet/test/wallet_tests.cpp:46
#7  0x0000555555ce03f2 in wallet_tests::CreateWallet::test_method (this=0x7fffffffc9c0) at wallet/test/wallet_tests.cpp:699
#8  0x0000555555cdf585 in wallet_tests::CreateWallet_invoker () at wallet/test/wallet_tests.cpp:696
```

ryanofsky commented:
Following fix seems to work:

```diff
--- a/src/wallet/test/init_test_fixture.cpp
+++ b/src/wallet/test/init_test_fixture.cpp
@@ -3,6 +3,7 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include <fs.h>
+#include <univalue.h>
 #include <util/check.h>
 #include <util/system.h>
 
@@ -37,6 +38,9 @@ InitWalletDirTestingSetup::InitWalletDirTestingSetup(const std::string& chainNam
 
 InitWalletDirTestingSetup::~InitWalletDirTestingSetup()
 {
+    gArgs.LockSettings([&](util::Settings& settings) {
+        settings.forced_settings.erase("walletdir");
+    });
     fs::current_path(m_cwd);
 }
 
```

Would suggest commit message like:

> test: Clear forced -walletdir setting after wallet init_tests
>
> Leaving this value set interfered with the CreateWallet test if it happened to execute later in the test ordering. Specifically it would cause CreateWallet test to write data to the current directory instead of temporary test directory.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/20932#issue-554809577

> This PR is based on @ryanofsky suggestion here [#20744 (review)](https://github.com/bitcoin/bitcoin/pull/20744#pullrequestreview-558131849) and it should help make #20744 diff smaller in the long run.

Would suggest writing a self-contained PR description that doesn't require reading a comment in another issue to understand. Something like:

refactor: Replace fs::absolute calls with AbsPathJoin calls

This adds better test coverage and will make it easier in #20744 to remove our dependency on the two-argument `boost::filesystem::absolute()` function which does not have a direct equivalent in C++17.

This PR doesn't change behavior aside from adding an assert and fixing a test bug.

kiminuo committed:
test: Clear forced -walletdir setting after wallet init_tests

Leaving this value set interfered with the CreateWallet test if it happened to execute later in the test ordering. Specifically it would cause CreateWallet test to write data to the current directory instead of temporary test directory.

kiminuo commented on src/fs.h:
```diff
@@ -21,6 +21,15 @@ namespace fs = boost::filesystem;
+     * @param[in] p     Path to combine with \p base.
+     * @return \p p if \p p is an absolute path, otherwise \p base joined with \p p is returned.
+     */
+    fs::path AbsPathJoin(const fs::path& base, const fs::path& p);
```
Yes, thanks.

ryanofsky commented on src/test/fs_tests.cpp:
```diff
@@ -52,6 +52,25 @@ BOOST_AUTO_TEST_CASE(fsbridge_fstream)
+        BOOST_CHECK_EQUAL(tmpfile2, p);
+    }
+    {
+        // Joining with empty path behaves differently in various filesystem implementations.
+        fs::path p = fsbridge::AbsPathJoin(tmpfolder, "");
+        BOOST_CHECK(p.is_absolute());
+        // "tmpfolder" does not contain trailing slash. No trailing slash should be added by the tested method.
+        BOOST_CHECK(p.string().back() != '/');
```
In commit "Replace fs::absolute calls with AbsPathJoin calls" (215c78465059175505218566f2ec38d52d3be827)

I don't think the comment and test are quite right. Joining behavior is well defined for boost and std path types and does not depend on the filesystem. The differences are between the boost library behavior and the standard library behavior are design choices, not OS differences. Would suggest replacing with:

```c++
{
     // Ensure joining with empty paths does not add trailing path components.
     BOOST_CHECK_EQUAL(tmpfile1, fsbridge::AbsPathJoin(tmpfile1, ""));
     BOOST_CHECK_EQUAL(tmpfile1, fsbridge::AbsPathJoin(tmpfile1, {}));
}
```

This is the empty path behavior that we are relying on AbsPathJoin and boost::filesystem::absolute functions for that the C++17 standard library doesn't directly provide in https://en.cppreference.com/w/cpp/filesystem/path/append

kiminuo replied:
Yes, it's bad wording on my side. Thanks for the suggestion.

ryanofsky reviewed (approved):
Code review ACK 215c78465059175505218566f2ec38d52d3be827. Thanks for updates!

felipsoarez reviewed (commented):
Concept ACK

ryanofsky commented on src/fs.h:
```diff
@@ -21,6 +21,15 @@ namespace fs = boost::filesystem;
+     *
+     * @param[in] base  Absolute path
+     * @param[in] path  Path to combine with base
+     * @returns path if it is an absolute path, otherwise base joined with path
```
In commit "Replace fs::absolute calls with AbsPathJoin calls" (215c78465059175505218566f2ec38d52d3be827)

Maybe expand this line a little to say "@returns path unchanged if it is an absolute path, otherwise returns base joined with path. Returns base unchanged if path is empty."

If desired, could also add preconditions and postconditions (base path must be absolute, returned path will always be absolute).

The information about paths being unchanged could be useful to callers and would be hard to discover without digging into the implementation and library documentation. It also explains how this join is different than other possible joins.

jonatack commented on src/fs.h:
```diff
@@ -21,6 +21,15 @@ namespace fs = boost::filesystem;
+    /**
+     * Helper function for joining paths
+     *
+     * @param[in] base  Absolute path
```
```suggestion
     * @param[in] base  Base path (must be absolute)
```

jonatack replied:
better yet, see the Doxygen section in doc/developer-notes.md and add
```
 * @pre  Base must be absolute
 * @post Returned path will always be absolute
```

jonatack commented on src/fs.h:
```diff
@@ -21,6 +21,15 @@ namespace fs = boost::filesystem;
     FILE *fopen(const fs::path& p, const char *mode);
 
+    /**
+     * Helper function for joining paths
```
```suggestion
     * Helper function for joining two paths to return an absolute path
```

jonatack reviewed (commented):
ACK d867d7a64c5c modulo two comments based on ryanofsky's excellent feedback in https://github.com/bitcoin/bitcoin/pull/20932#discussion_r558485716 "If desired, could also add preconditions and postconditions (base path must be absolute, returned path will always be absolute)."

ryanofsky reviewed (approved):
Code review ACK da9caa1cedd69702aea44cb44b2fd0a2d6d56916. Just comment and test tweaks since previous review.

Final change is very simple and easy to review, and I think makes code more understandable and improves tests. Thanks for implementing this!

jonatack commented:
Code review ACK da9caa1cedd69702aea44cb44b2fd0a2d6d56916 only doxygen improvements since my last review per `git diff d867d7a da9caa1`

Thanks for updating.

MarcoFalke commented:
review ACK da9caa1cedd69702aea44cb44b2fd0a2d6d56916 📯

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK da9caa1cedd69702aea44cb44b2fd0a2d6d56916 📯
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgRdQv8DzNpEk8/k2bFwErjQdb2bW93pR3nNfNIarBpkEE+QHpmP7vSlMXz+04q
JlOLmdAPEVv2jWA+d2lEdvoMapyCEBbUvii7jSgxBN/Qb6ONyXddjRTjDEVXmJZW
UCjgJ9hfo59NxuA49LcAVPMVatdkA7PtgKkYZ78sG14BlJ+15csFYA8R+FzMW+hh
kD/j8tYTQQQ1KN11Vmyw08rLcUU/XFavbCf/0++FlDGKbZOROYVtfhqxybyn4YzN
Ibt/AJq+dGIOBB3LY9cN0Muwb8sxVZQrM4GWxNUwHzkH1VHZL7vdVGGWJBx+rU2c
5mICi+9n60wmeRUahHH5hAT2+hr5XDgBo1i5mZUpZYmTZXp+kO2a83LI/oNGpkK5
MTWy4K+5koCzgLPQNDRGRiVM/aB81F0l4LsaY9/q5ANDQg5LP6YWwyHnuzbLujHs
pHTn2rUe+DnLmEM94a6p05Ht0NsfyWusw/NpTkAYl9ZkvNjqixRmIbkwIm7sYC65
UrqATWdd
=3DWR
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `635e482e0a7be99ddac7857318934ace812fbbbfd5d8237d0445be8d5d77dff1  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108635e482e0a7be99ddac7857318934ace812fbbbfd5d8237d0445be8d5d77dff1f010ee8d02f62d80ad44006beb6c1593d46308fff0100a923c693cd2de2653e1d3a68c52253708f1046009be27f00881850a2fbd8b412b0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01078f94922ce46b6e71899b5de550212d408f1046009be26f00854babc00733ff3490083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010cc81ff3f94be6d764cf31f0ea3a00bab08f1046009be27f008a99169ecc4975c8d0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010fb51367f1075f1f064404b6e8e3a12f508f1046009be27f00844bf618bdd36e0320083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>
